mypy/build.py

```diff
@@ -778,95 +785,188 @@ def is_file(path: str) -> bool:
+            return None
+
+        # We can add at most one module implementation to paths
+        # But module can have multiple stubs
+        py_path = path.endswith('.py')
+        if self.has_py and py_path:
+            return None
+
+        if type == ModuleType.namespace:
+            ok = self._verify_namespace(path)
+        else:
+            ok = self._verify_module(path)
+
+        if not ok:
+            return None
+
+        if py_path:
+            self.has_py = True
+
+        self.type = type
+        self.paths.append(path)
+
+    def _verify_module(self, path: str) -> bool:
+        # At this point we already know that that it's valid python path
+        # We only need to check file existence
+        if not is_file(path):
             return False
-    return True
+
+        return True
+
+    def _verify_namespace(self, path: str) -> bool:
+        if os.path.isdir(path):
+            files = set(list_dir(path) or [])  # type: Set[str]
+            if not ('__init__.py' in files or '__init__.pyi' in files):
+                return True
+
+        return False
+
+
+def is_pkg_path(path: str) -> bool:
+    return path.endswith(('__init__.py', '__init__.pyi'))
+
+
+def is_module_path(path: str) -> bool:
+    return not is_pkg_path(path) and path.endswith(('.py', '.pyi'))
+
+
+def is_namespace_path(path: str) -> bool:
+    return path.endswith(os.path.sep)
+
+
+class ModuleDiscovery:
+    def __init__(self,
+                 lib_path: Iterable[str],
+                 namespaces_allowed: bool = False) -> None:
+
+        self.lib_path = tuple(os.path.normpath(p) for p in lib_path)  # type: Tuple[str, ...]
+        self.namespaces_allowed = namespaces_allowed
+        self._find_module_cache = {}  # type: Dict[str, List[str]]
+
+    def find_module(self, id: str) -> Optional[str]:
+        paths = self._find_module(id)
+        if paths:
+            return paths[0]
+        else:
+            return None
+
+    def find_modules_recursive(self, module: str) -> List[BuildSource]:
+        """
+        Discover module and all it's children
+        Remove duplicates from discovered paths
+        """
+        hits = set()  # type: Set[str]
+        result = []  # type: List[BuildSource]
+        for src in self._find_modules_recursive(module):
+            if src.module not in hits:
+                hits.add(src.module)
+                result.append(src)
+        return result
+
+    def _find_modules_recursive(self, module: str) -> List[BuildSource]:
+        module_paths = self._find_module(module)
+
+        srcs = []  # type: List[BuildSource]
+        for path in module_paths:
+            if is_module_path(path) or is_pkg_path(path):
+                srcs.append(BuildSource(path, module, None))
+
+            if is_pkg_path(path):
+                path = dirname(path)
+                for submodule in self._find_submodules(module, path):
+                    srcs += self._find_modules_recursive(submodule)
```

Comment:
by going all the way back to `_find_module` (which starts at the very beginning with all of `lib_path`) on every recursive call here, we are going to do a lot of repeated unnecessary work. It's worse with namespace packages than with the previous code. Imagine a namespace package that exists in three different paths, `dir1/foo`, `dir2/foo`, and `dir3/foo`. Consider that each of those has a different sub-package, `dir1/foo/a/`, `dir2/foo/b`, and `dir3/foo/c`. When we find `dir1/foo/a`, we throw away the context we already have (that we are within `dir1/foo/a`) and just call `_find_module('foo.a')`, which will end up unnecessarily looking at every single `lib_path` entry again to see if it has a `foo/a/` within it.

It may be that caching can reduce the impact of this multiplied repeated work, but not fully. And I think we can easily do better in the algorithm. `_collect_paths` already has the right logic to look for the final component of a submodule within a given limited set of parent paths. Can't we use `_collect_paths` here instead of recursive calls to be more efficient and only look within the relevant parent?